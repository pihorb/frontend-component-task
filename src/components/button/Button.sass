.button
  background-color: #FA6B30
  border: none
  color: white
  font-weight: bold
  padding: .7em 1em
  font-size: 1em
  cursor: pointer
  min-width: 15%
  opacity: .9
  transition: opacity .5s
  margin-top: 2em
  &:hover
    opacity: 1
