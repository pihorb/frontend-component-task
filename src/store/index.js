export {context} from './context';
export {reducer} from './reducer';
export {types} from './types';
