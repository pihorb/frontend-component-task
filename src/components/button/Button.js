import React from 'react';
import './Button.sass';
import {useAppContext} from '../../context/AppContext';

const Button = ({children}) => {
  const {role, checkBoxes, lastChoice} = useAppContext();
  const state = {
    currnetRole: role,
    lastSelected: lastChoice,
    ...checkBoxes
  };

  return (
    <button className="button" onClick={() => console.log(state)}>
      {children}
    </button>
  );
};

export default Button;
