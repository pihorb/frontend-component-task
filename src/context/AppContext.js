import React, {createContext, useContext, useReducer} from 'react';
import reducer from '../reducers/reducer';
import {ADD_FOLDERS, ADD_GEMS, REMOVE_FOLDERS, REMOVE_GEMS} from '../reducers/types';

const AppContext = createContext();

export const useAppContext = () => {
  return useContext(AppContext);
};

export const setCheckBoxes = (...opts) => {
  const min = ['view', 'share'];

  if (opts) min.push(...opts);

  return min;
};

export const AppProvider = ({children}) => {
  const [state, dispatch] = useReducer(reducer, {
    role: 'user',
    lastChoice: null,
    checkBoxes: {
      gems: setCheckBoxes('create', 'update'),
      folders: setCheckBoxes('create', 'update')
    }
  });

  const setActiveRole = (role) => {
    dispatch({type: role.name.toUpperCase()});
  };

  const updateCheckBoxes = (name, path, isChecked) => {
    switch (true) {
      case isChecked && path === 'gems':
        return dispatch({type: REMOVE_GEMS, payload: name});
      case isChecked && path === 'folders':
        return dispatch({type: REMOVE_FOLDERS, payload: name});
      case path === 'folders':
        return dispatch({type: ADD_FOLDERS, payload: name});
      case path === 'gems':
        return dispatch({type: ADD_GEMS, payload: name});
      default:
        return null;
    }
  };

  return (
    <AppContext.Provider
      value={{
        toggle: setActiveRole,
        updateCheckBoxes,
        role: state.role,
        checkBoxes: state.checkBoxes,
        lastChoice: state.lastChoice
      }}
    >
      {children}
    </AppContext.Provider>
  );
};
