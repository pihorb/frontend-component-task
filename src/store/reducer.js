import {
  ADD_FOLDERS,
  ADD_GEMS,
  ADMIN,
  CUSTOM,
  MEMBER,
  REMOVE_FOLDERS,
  REMOVE_GEMS,
  USER
} from './types';
import {defaultSelected} from './context';

const reducer = (state, action) => {
  switch (action.type) {
    case USER:
      return {
        ...state,
        role: 'user',
        slots: {
          gems: defaultSelected('create', 'update'),
          folders: defaultSelected('create', 'update')
        }
      };
    case MEMBER:
      return {
        ...state,
        role: 'member',
        slots: {
          gems: defaultSelected(),
          folders: defaultSelected()
        }
      };
    case ADMIN:
      return {
        ...state,
        role: 'admin',
        slots: {
          gems: defaultSelected('create', 'update', 'move', 'delete'),
          folders: defaultSelected('create', 'update', 'move', 'delete')
        }
      };
    case CUSTOM:
      return {
        ...state,
        role: 'custom',
        slots: {
          gems: [],
          folders: []
        }
      };
    case ADD_GEMS:
      return {
        ...state,
        lastChoice: action.payload,
        slots: {
          gems: [...state.slots.gems, action.payload],
          folders: [...state.slots.folders]
        }
      };
    case ADD_FOLDERS:
      return {
        ...state,
        lastChoice: action.payload,
        slots: {
          gems: [...state.slots.gems],
          folders: [...state.slots.folders, action.payload]
        }
      };
    case REMOVE_GEMS:
      return {
        ...state,
        slots: {
          gems: state.slots.gems.filter((item) => item !== action.payload),
          folders: [...state.slots.folders]
        }
      };
    case REMOVE_FOLDERS:
      return {
        ...state,
        slots: {
          gems: [...state.slots.gems],
          folders: state.slots.folders.filter((item) => item !== action.payload)
        }
      };
    default:
      return state;
  }
};

export default reducer;
