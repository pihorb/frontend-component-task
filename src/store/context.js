import React, {createContext, useContext, useReducer} from 'react';
import reducer from './reducer';
import {ADD_FOLDERS, ADD_GEMS, REMOVE_FOLDERS, REMOVE_GEMS} from './types';

const AppContext = createContext();

export const useAppContext = () => {
  return useContext(AppContext);
};

export const defaultSelected = (...opts) => {
  const min = ['view', 'share'];

  if (opts) min.push(...opts);

  return min;
};

export const AppProvider = ({children}) => {
  const [state, dispatch] = useReducer(reducer, {
    role: 'user',
    lastChoice: null,
    slots: {
      gems: defaultSelected('create', 'update'),
      folders: defaultSelected('create', 'update')
    }
  });

  const toggleRole = (role) => {
    dispatch({type: role.name.toUpperCase()});
  };

  const toggleSelect = (name, path, isChecked) => {
    switch (true) {
      case isChecked && path === 'gems':
        return dispatch({type: REMOVE_GEMS, payload: name});
      case isChecked && path === 'folders':
        return dispatch({type: REMOVE_FOLDERS, payload: name});
      case path === 'folders':
        return dispatch({type: ADD_FOLDERS, payload: name});
      case path === 'gems':
        return dispatch({type: ADD_GEMS, payload: name});
      default:
        return null;
    }
  };

  return (
    <AppContext.Provider
      value={{
        toggleRole,
        toggleSelect,
        role: state.role,
        slots: state.slots,
        lastChoice: state.lastChoice
      }}
    >
      {children}
    </AppContext.Provider>
  );
};
