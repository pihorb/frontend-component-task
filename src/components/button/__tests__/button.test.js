import renderer from 'react-test-renderer';
import '@testing-library/jest-dom';
import React from 'react';
import {cleanup, render} from '@testing-library/react';
import Button from '../Button';
import {AppProvider} from '../../../store/context';

const renderButton = (cb, children) => {
  return cb(
    <AppProvider>
      <Button>{children}</Button>
    </AppProvider>
  );
};

beforeEach(cleanup);

describe('button component test', () => {
  it('renders correctly', () => {
    const {queryByTestId} = renderButton(render);

    expect(queryByTestId('btn')).toBeTruthy();
  });

  it('should have class button', () => {
    const {queryByTestId} = renderButton(render, 'click me!');

    expect(queryByTestId('btn').classList.contains('button')).toBe(true);
  });

  it('should display child', () => {
    const {queryByTestId} = renderButton(render, 'click me!');

    expect(queryByTestId('btn')).toHaveTextContent('click me!');
  });

  it('check if snapshot is matching', () => {
    const tree = renderButton(renderer.create).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
