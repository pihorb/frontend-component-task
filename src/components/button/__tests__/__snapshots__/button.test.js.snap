// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component test check if snapshot is matching  1`] = `
<button
  className="button"
  data-testid="btn"
  onClick={[Function]}
/>
`;
